File: openrag/api/routers/admin/jobs.py
"""Queue routes — thin HTTP layer over :class:`JobService`.

Phase 8D.2: queue aggregation and the ``?task_status=`` filtering moved
to ``services.orchestrators.job_service.JobService``. This module keeps
HTTP transport only: the shared admin/current-user ``Depends`` wrappers
and the ``request.url_for`` link building.
"""

from api.dependencies.auth import current_user, require_admin
from di.providers import get_job_service
from fastapi import APIRouter, Depends, Request, status
from fastapi.responses import JSONResponse

router = APIRouter()


@router.get(
    "/info",
    description="""Get queue and worker pool information.

**Permissions:**
- Requires admin role

**Response:**
Returns system status including:

**Workers:**
- `total_slots`: Total available worker capacity
- `pool_size`: Number of worker actors
- `max_per_actor`: Max concurrent tasks per worker

**Tasks:**
- `active`: Total active tasks
- `active_statuses`: Breakdown by status (QUEUED, SERIALIZING)
- `total_completed`: Count of completed tasks
- `total_cancelled`: Count of cancelled tasks
- `total_failed`: Count of failed tasks

**Use Case:**
Monitor system load and worker utilization.
""",
)
async def get_queue_info(
    admin=Depends(require_admin),
    service=Depends(get_job_service),
):
    return await service.get_queue_info()


@router.get(
    "/tasks",
    name="list_tasks",
    description="""List indexing tasks with optional filtering.

**Query Parameters:**
- `task_status`: Filter by status (optional)
  - `active`: Show QUEUED or SERIALIZING tasks
  - `completed`: Show completed tasks
  - `failed`: Show failed tasks
  - `cancelled`: Show cancelled tasks
  - Any exact status name (case-insensitive)
  - Omit to show all tasks

**Permissions:**
- Regular users: See only their own tasks
- Admins: See all tasks

**Response:**
Returns list of tasks with:
- `task_id`: Unique task identifier
- `state`: Current task state
- `details`: Task metadata (file_id, partition, etc.)
- `created_at`: Date and time when the task was created
- `duration_ms`: Total task duration in milliseconds
- `url`: Link to detailed task status
- `error_url`: Link to error details (if failed)

**Task States:**
- `QUEUED`: Waiting to start
- `SERIALIZING`: Parsing, chunking, embedding, and storing the document
- `COMPLETED`: Successfully finished
- `CANCELLED`: Cancelled by user/admin
- `FAILED`: Error occurred
""",
)
async def list_tasks(
    request: Request,
    task_status: str | None = None,
    user=Depends(current_user),
    service=Depends(get_job_service),
):
    """
    - ?task_status=active  → QUEUED | SERIALIZING
    - ?task_status=<exact> → exact match (case-insensitive)
    - (none)               → all tasks
    """
    rows = await service.list_tasks(
        is_admin=bool(user.get("is_admin")),
        user_id=user.get("id"),
        task_status=task_status,
    )

    tasks = []
    for row in rows:
        task_id = row["task_id"]
        item = {
            "task_id": task_id,
            "state": row["state"],
            "details": row["details"],
            "created_at": row["created_at"],
            "duration_ms": row["duration_ms"],
            **(
                {"error_url": str(request.url_for("get_task_error", task_id=task_id))}
                if row["state"] == "FAILED"
                else {}
            ),
            "url": str(request.url_for("get_task_status", task_id=task_id)),
        }
        tasks.append(item)

    return JSONResponse(status_code=status.HTTP_200_OK, content={"tasks": tasks})

File: ui/src/lib/api/jobs.ts
import { request } from "./client";

// OpenRag jobs/queue API (poll-based — OpenRag has NO task SSE).
// Verified vs routers/admin/jobs.py + routers/admin/indexing.py. The jobs
// pages poll getTaskStatus (+ getTaskLogs) on a refetchInterval, stopping
// once isTerminalState(state).
//   GET    /queue/info                          queue + worker pool summary
//   GET    /queue/tasks[?task_status=]          list tasks → { tasks: [...] }
//   GET    /indexer/task/{id}                   status → { task_id, task_state, details, error_url? }
//   GET    /indexer/task/{id}/error             { task_id, traceback: string[] }
//   GET    /indexer/task/{id}/logs[?max_lines]  { task_id, logs: string[] }
//   DELETE /indexer/task/{id}                   cancel → { message }

const QUEUE = "/queue";
const TASK = "/indexer/task";

export type TaskState =
  | "QUEUED"
  | "SERIALIZING"
  | "CHUNKING"
  | "INSERTING"
  | "COMPLETED"
  | "FAILED"
  | "CANCELLED";

const ACTIVE_STATES: readonly TaskState[] = ["QUEUED", "SERIALIZING", "CHUNKING", "INSERTING"];
const TERMINAL_STATES: readonly TaskState[] = ["COMPLETED", "FAILED", "CANCELLED"];

export const isActiveState = (s: string): boolean => (ACTIVE_STATES as readonly string[]).includes(s);
/** True once polling can stop — the task won't change state again. */
export const isTerminalState = (s: string): boolean => (TERMINAL_STATES as readonly string[]).includes(s);

export interface TaskDetails {
  file_id: string;
  partition: string;
  metadata: Record<string, unknown>;
  user_id: number;
  failed_stage?: string;
}

/** Row from GET /queue/tasks — note `state` (vs `task_state` in the detail). */
export interface TaskListItem {
  task_id: string;
  state: TaskState;
  details: TaskDetails;
  created_at?: string | null;
  duration_ms?: number | null;
  url: string;
  error_url?: string;
}

/** GET /indexer/task/{id} — note `task_state` (vs `state` in list rows). */
export interface TaskStatus {
  task_id: string;
  task_state: TaskState;
  details: TaskDetails;
  error_url?: string;
}

export interface QueueInfo {
  workers: { total_slots: number; pool_size: number; max_per_actor: number };
  tasks: {
    active: number;
    active_statuses: Record<string, number>;
    total_completed: number;
    total_cancelled: number;
    total_failed: number;
  };
}

export function getQueueInfo(): Promise<QueueInfo> {
  return request<QueueInfo>(`${QUEUE}/info`);
}

export function listTasks(taskStatus?: string): Promise<{ tasks: TaskListItem[] }> {
  const qs = taskStatus ? `?task_status=${encodeURIComponent(taskStatus)}` : "";
  return request<{ tasks: TaskListItem[] }>(`${QUEUE}/tasks${qs}`);
}

export function getTaskStatus(taskId: string): Promise<TaskStatus> {
  return request<TaskStatus>(`${TASK}/${encodeURIComponent(taskId)}`);
}

export function getTaskError(taskId: string): Promise<{ task_id: string; traceback: string[] }> {
  return request<{ task_id: string; traceback: string[] }>(`${TASK}/${encodeURIComponent(taskId)}/error`);
}

export function getTaskLogs(taskId: string, maxLines = 100): Promise<{ task_id: string; logs: string[] }> {
  return request<{ task_id: string; logs: string[] }>(`${TASK}/${encodeURIComponent(taskId)}/logs?max_lines=${maxLines}`);
}

export function cancelTask(taskId: string): Promise<{ message: string }> {
  return request<{ message: string }>(`${TASK}/${encodeURIComponent(taskId)}`, { method: "DELETE" });
}
